src/hostnet/lib/tcpip_stack.ml

```diff
@@ -29,37 +28,47 @@ let rec smallest_prefix a_ip other_ips = function
     then prefix
     else smallest_prefix a_ip other_ips (bits - 1)
 
+(* given some MACs and IPs, construct a usable DHCP configuration *)
 let make ~client_macaddr ~server_macaddr ~peer_ip ~local_ip =
-  (* FIXME: We need a third IP just to make the DHCP server happy *)
+  let open Dhcp_server.Config in
+  (* FIXME: We need a DHCP range to make the DHCP server happy, even though we
+     intend only to serve IPs to one downstream host.
+     see https://github.com/haesbaert/charrua-core/issues/27 - this may be
+     resolved in the future *)
   let low_ip, high_ip =
     let open Ipaddr.V4 in
     let highest = if compare local_ip peer_ip < 0 then peer_ip else local_ip in
     let i32 = to_int32 highest in
     of_int32 @@ Int32.succ i32, of_int32 @@ Int32.succ @@ Int32.succ i32 in
   let prefix = smallest_prefix peer_ip [ local_ip; low_ip; high_ip ] 32 in
-  { local_ip; peer_ip; low_ip; high_ip; prefix; client_macaddr; server_macaddr }
-
-let dhcp_conf ~config =
-  let network = Ipaddr.V4.(to_string (Prefix.network config.prefix)) in
-  let netmask = Ipaddr.V4.(to_string (Prefix.netmask config.prefix)) in
-  let local_ip = Ipaddr.V4.to_string config.local_ip in
-  let peer_ip = Ipaddr.V4.to_string config.peer_ip in
-  let low_ip = Ipaddr.V4.to_string config.low_ip in
-  let high_ip = Ipaddr.V4.to_string config.high_ip in
-
-  Printf.sprintf "\n\
-  option domain-name \"local\";\n\
-  subnet %s netmask %s {\n\
-    option routers %s;\n\
-    option domain-name-servers %s;\n\
-    range %s %s;\n\
-    host xhyve {\n\
-      hardware ethernet %s;\n\
-      fixed-address %s;\n\
-    }\n\
-  }\n\
-  " network netmask local_ip local_ip low_ip high_ip
-  (Macaddr.to_string config.client_macaddr) peer_ip
+  (* 
+  let config ~config =
+    Dhcp_server.Config.parse (dhcp_conf ~config) [(config.local_ip,
+    config.server_macaddr)] *)
+  let options = [
+    Dhcp_wire.Domain_name "local";
+    Dhcp_wire.Routers [ local_ip ];
+    Dhcp_wire.Name_servers [ local_ip ];
+    Dhcp_wire.Time_servers [ local_ip ];
+  ] in
+  let hyperkit : host = {
+    hostname = "hyperkit";
+    options = [];
+    hw_addr = Some client_macaddr;
+    fixed_addr = Some peer_ip;
+  } in
+  let dhcp_configuration : Dhcp_server.Config.t = {
+    options = options;
+    hostname = "vpnkit"; (* it's us! *)
+    hosts = [ hyperkit ];
+    default_lease_time = Int32.of_int (60 * 60 * 2); (* 2 hours, from charrua defaults *)
+    max_lease_time = Int32.of_int (60 * 60 * 24) ; (* 24 hours, from charrua defaults *)
+    ip_addr = local_ip;
+    mac_addr = server_macaddr;
+    network = prefix;
+    range = (low_ip, high_ip);
+  } in
+  { peer_ip; local_ip; prefix; server_macaddr; client_macaddr; dhcp_configuration }
```

Comment:
That's looking a lot nicer!

BTW do you know if charrua allows the dynamic range to be empty? I tried various things with the string and failed to make it work.